serverless/deploy/github/combine_witness_signatures/combine_witness_signatures.go

```diff
@@ -0,0 +1,209 @@
+
+// merge_witness_signatures is a tool to manage the checkpoint.witnessed file.
+//
+// Given the following files:
+// - checkpoint: a log checkpoint
+// - checkpoint.witnessed: a log checkpoint with additional witness signatures (optional)
+// - a collection of additional witnessed checkpoints
+// this tool will determine whether to update an existing checkpoint.witnessed
+// from the additional witnessed checkpoints, or replace it entirely by merging
+// witness cosignatures with the contents of the checkepoint file.
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/checkpoints"
+	"golang.org/x/mod/sumdb/note"
+)
+
+var (
+	storageDir     = flag.String("storage_dir", "", "Root directory of the log.")
+	logKey         = flag.String("log_public_key", "", "Log's public key.")
+	witnessKeys    = flag.String("witness_public_key_files", "", "One or more space-separated globs matching all the known witness keys.")
+	numRequired    = flag.Uint("required_witness_sigs", 0, "Specifies the minimum number of valid witness signatures required to be present on the output to consider the merge successful.")
+	output         = flag.String("output", "", "Output file to write combined checkpoint to.")
+	deleteConsumed = flag.Bool("delete_consumed", true, "Whether to delete files containing merged checkpoints.")
+)
+
+func main() {
+	flag.Parse()
+
+	logSigV, err := note.NewVerifier(*logKey)
+	if err != nil {
+		glog.Exitf("Failed to parse log public key: %v", err)
+	}
+
+	witSigV, err := witnessVerifiers(*witnessKeys)
+	if err != nil {
+		glog.Exitf("Failed to load witness keys: %v", err)
+	}
+
+	cpNote, cpRaw, err := openNote(filepath.Join(*storageDir, "checkpoint"), note.VerifierList(logSigV))
+	if err != nil {
+		glog.Exitf("Failed to open checkpoint: %v", err)
+	}
+
+	cpwNote, cpwRaw, err := openNote(filepath.Join(*storageDir, "checkpoint.witnessed"), witSigV)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			glog.Exitf("Failed to open checkpoint.witnessed: %v", err)
+		}
+	}
+
+	witnessDir := filepath.Join(*storageDir, "witness")
+
+	ans, err := additionalNotes(filepath.Join(witnessDir, "*"))
+	if err != nil {
+		glog.Exitf("Failed to read additional checkpoints: %v", err)
+	}
+
+	var out []byte
+	// If checkpoints are different (or there is no checkpoint.witnessed file), see
+	// if we have sufficient witness cosigs to promote checkpoint to
+	// checkpoint.witnessed
+	if cpwNote == nil || cpNote.Text != cpwNote.Text {
+		out, err = mergeCheckpoints(cpRaw, ans, *numRequired, logSigV, witSigV, false)
+		if err != nil {
+			glog.Warningf("Failed to merge with new checkpoint file: %v", err)
+		} else {
+			// Since we've just promoted the new checkpoint and merged any compatible
+			// cosignatures, we can delete all additional Checkpoints:
+			if err := os.RemoveAll(witnessDir); err != nil {
```

Comment:
Should this be guarded by `delete_consumed` too? That flag seems a little niche otherwise. With it also guarding this behaviour its behaviour is far simpler to understand.

Reply:
This is a good question - the thinking here is that since we've just promoted (and combined any possible cosigs)`checkpoint` there can be nothing useful left in the `witness` directory, so we delete *everything* including non-consumed.

In the other branch, we only delete the consumed files because, conversely, there *may* be some number of useful cosigs over `checkpoint` there but which are insufficient to meet the threshold for promoting `checkpoint` to `checkpoint.witnessed`.

I've added a comment explaining the absence of deletion in the "cosign checkpoint.witnessed" branch.